In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observation
<br>
Initial observations include the following:
 - The columns are not named according to the python standard.The column names use camelcase instead of python's snakecase.
 - Some of the colums contain null values.
 - The dataset contains 20 columns, majority of the columns are string
 - Symbols are included in the values.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename({"dateCrawled":"date_crawled", 
              "offerType":"offer_type", 
              "vehicleType":"vehicle_type", 
              "yearOfRegistration":"registration_year", 
              "powerPS":"power_ps", 
              "monthOfRegistration":"registration_month", 
              "fuelType": "fuel_type" ,
              "notRepairedDamage":"unrepaired_damage", 
              "dateCreated":"ad_created", 
              "nrOfPictures": "pictures_nr", 
              "postalCode":"postal_code", 
              "lastSeen": "last_seen"}, axis=1, inplace=True)

In [7]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_nr', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  pictures_nr  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I have changed the column names from camelcase to snakecase and assigned them back to the autos dataset, so as to make them more descriptive and readable. 

In [9]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_nr   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

The following observations were made:
- There are columns which have one value and need to be dropped.
- The pictures_nr column mostly has one value
- Postal code column
- seller column
- unrepaired damage
- The symbol added to the price column as well as the odometer column needs to be cleaned and stored as float or integer.

In [10]:
#remove irerelevant columns
autos.drop(columns = ["seller", "pictures_nr", "offer_type"], inplace=True)

In [11]:
#Removing the non-numeric characters in the price columns
autos["price"] = (autos["price"]
                .str.replace("$", "")
                .str.replace(",", "")
                .astype(float)
                )
#Rename the price column to make it more descriptive
autos.rename({"price": "price_usd"}, axis=1, inplace=True)


In [12]:
#Removing the non-numeric charcters in the odometer column
autos["odometer"] = (autos["odometer"]
                   .str.replace(",", "")
                   .str.replace("km", "")
                   .astype(float)
                   )
            
#Rename the odometer column to make it more descriptive
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

## Exploring the price_usd column

In [13]:
#exploring the data: exploring the price column
print(autos["price_usd"].unique().shape)
autos["price_usd"].describe()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [14]:
#Determine the top 20 prices
top_20_prices = autos["price_usd"].value_counts().sort_index(ascending=False).head(20)
print(f"Top 20 prices:\n{top_20_prices}")


Top 20 prices:
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64


In [15]:
#Determine the least 20 prices
least_20_prices = autos["price_usd"].value_counts().sort_index(ascending=True).head(20)
print(f"Least 20 prices:\n{least_20_prices}")

Least 20 prices:
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64


In [16]:
autos.loc[autos["price_usd"] == 1.0]

date_crawled                                               name  \
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
344    2016-03-22 11:25:18  Verkaufe_hier_mein_mein_schoener_honda_crx_in_...   
791    2016-04-05 09:52:11                              Opel_Vectra_in_teilen   
981    2016-03-10 00:56:12                            Verkaufen_oder_tauschen   
1115   2016-03-24 22:36:58                           Suche_Passat_VR6_Variant   
...                    ...                                                ...   
49240  2016-03-10 09:54:48   Suche_7_Sitzer_Espace_oder?_Biete_s.Beschreibung   
49377  2016-03-20 00:58:36             Tausche_BMW_318i_Facelift_Model_143_ps   
49528  2016-03-28 19:57:41       BMW_E36__316i_Compact_M_Paket_"Schlachtfest"   
49728  2016-03-25 16:00:11                     Ford_Focus_1.4_16V_Bastlerauto   
49845  2016-03-21 00:50:15  Schlachte_VW_Sharan_vr6_Automatik___no_GTI_16V...   

       price_usd   abtest vehicle_type  registration_year    gearbox  \
55           1.0     test          NaN               2017  automatik   
344          1.0     test          NaN               2000    manuell   
791          1.0  control        kombi               1999    manuell   
981          1.0  control        kombi               2003    manuell   
1115         1.0  control          NaN               1995    manuell   
...          ...      ...          ...                ...        ...   
49240        1.0     test        kombi               1999    manuell   
49377        1.0  control    limousine               2002    manuell   
49528        1.0  control          NaN               2016    manuell   
49728        1.0     test          NaN               2016    manuell   
49845        1.0     test          NaN               2000  automatik   

       power_ps     model  odometer_km  registration_month fuel_type  \
55          224  e_klasse     125000.0                   7    benzin   
344         125    andere       5000.0                   0       NaN   
791         101    vectra     150000.0                   1    benzin   
981         150       3er     150000.0                   9    diesel   
1115          0    passat     150000.0                   0    benzin   
...         ...       ...          ...                 ...       ...   
49240         0       NaN     150000.0                   4    benzin   
49377       143       3er     150000.0                   8    benzin   
49528       102       3er     150000.0                  12    benzin   
49728         0     focus     150000.0                   0    benzin   
49845       174    sharan       5000.0                   6    benzin   

               brand unrepaired_damage           ad_created  postal_code  \
55     mercedes_benz              nein  2016-03-06 00:00:00        22111   
344            honda               NaN  2016-03-22 00:00:00        65391   
791             opel               NaN  2016-04-05 00:00:00        59394   
981              bmw              nein  2016-03-10 00:00:00        58840   
1115      volkswagen               NaN  2016-03-24 00:00:00        38556   
...              ...               ...                  ...          ...   
49240            bmw               NaN  2016-03-10 00:00:00        41199   
49377            bmw               NaN  2016-03-19 00:00:00        86356   
49528            bmw               NaN  2016-03-28 00:00:00        49696   
49728           ford                ja  2016-03-25 00:00:00        34587   
49845     volkswagen              nein  2016-03-20 00:00:00        14621   

                 last_seen  
55     2016-03-08 05:45:44  
344    2016-04-06 18:15:57  
791    2016-04-07 13:16:37  
981    2016-04-06 01:16:59  
1115   2016-04-06 03:45:42  
...                    ...  
49240  2016-03-28 10:15:50  
49377  2016-03-20 04:56:34  
49528  2016-04-07 02:47:04  
49728  2016-03-25 16:40:21  
49845  2016-04-05 21:16:16  

[156 rows x 17 columns]

From the initial data exploration of the price column we can see  that the maximum price for a car is 10 000 000USD which is a lot for a used car while the minimum price for a car is 0USD which is impossible. In the data, we have about 1421 instances have a price of 0USD and 14 instances with prices above 350000USD
<br>
It shows that there are erros in the price_usd column and it needs further cleaning. 


## Removing Price Outliers
<br>
Prices lower than 50 USD and higher than 350 000USD will be removed.

In [17]:
#Removing price outliers
autos = autos[(autos["price_usd"] > 50.0) & (autos["price_usd"] < 350000.0)] 

## Exploring the odometer_km column

In [18]:
print(autos["odometer_km"].unique().shape)
autos["odometer_km"].describe()

(13,)


count     48297.000000
mean     125899.745326
std       39578.784414
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
#Determine the top 20 Odometer readings
top_20_odometer = autos["odometer_km"].value_counts().sort_index(ascending=False).head(20)
print(f"Top 20 Odometer readings:\n{top_20_odometer}")

Top 20 Odometer readings:
150000.0    31257
125000.0     5045
100000.0     2103
90000.0      1733
80000.0      1413
70000.0      1214
60000.0      1153
50000.0      1011
40000.0       815
30000.0       777
20000.0       757
10000.0       248
5000.0        771
Name: odometer_km, dtype: int64


In [20]:
least_20_odometer = autos["odometer_km"].value_counts().sort_index(ascending=True).head(20)
print(f"Least 20 Odometer readings:\n{least_20_odometer}")

Least 20 Odometer readings:
5000.0        771
10000.0       248
20000.0       757
30000.0       777
40000.0       815
50000.0      1011
60000.0      1153
70000.0      1214
80000.0      1413
90000.0      1733
100000.0     2103
125000.0     5045
150000.0    31257
Name: odometer_km, dtype: int64


The Odometer column looks good and does not need further cleaning

## Working with Date columns

In [21]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48297 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        48297 non-null  object 
 1   name                48297 non-null  object 
 2   price_usd           48297 non-null  float64
 3   abtest              48297 non-null  object 
 4   vehicle_type        43842 non-null  object 
 5   registration_year   48297 non-null  int64  
 6   gearbox             46061 non-null  object 
 7   power_ps            48297 non-null  int64  
 8   model               45889 non-null  object 
 9   odometer_km         48297 non-null  float64
 10  registration_month  48297 non-null  int64  
 11  fuel_type           44390 non-null  object 
 12  brand               48297 non-null  object 
 13  unrepaired_damage   39373 non-null  object 
 14  ad_created          48297 non-null  object 
 15  postal_code         48297 non-null  int64  
 16  last

There are 5 columns that represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
<br>
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
<br>
Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

In [22]:
#understanding how the values in the 3 strings columns are formatted
autos[["date_crawled", "ad_created", "last_seen"]] [0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
#Calculating the distribution of date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025364
2016-03-06    0.014080
2016-03-07    0.036006
2016-03-08    0.033253
2016-03-09    0.033046
2016-03-10    0.032238
2016-03-11    0.032569
2016-03-12    0.036980
2016-03-13    0.015674
2016-03-14    0.036648
2016-03-15    0.034288
2016-03-16    0.029464
2016-03-17    0.031534
2016-03-18    0.012879
2016-03-19    0.034723
2016-03-20    0.037787
2016-03-21    0.037269
2016-03-22    0.032921
2016-03-23    0.032300
2016-03-24    0.029464
2016-03-25    0.031534
2016-03-26    0.032279
2016-03-27    0.031079
2016-03-28    0.034930
2016-03-29    0.034122
2016-03-30    0.033687
2016-03-31    0.031865
2016-04-01    0.033750
2016-04-02    0.035572
2016-04-03    0.038553
2016-04-04    0.036524
2016-04-05    0.013065
2016-04-06    0.003168
2016-04-07    0.001387
Name: date_crawled, dtype: float64

For the date_crawled column, 2016-04-03 has the highest percentage of 3.855%

In [24]:
#Calculating the distribution of ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038802
2016-04-04    0.036897
2016-04-05    0.011781
2016-04-06    0.003251
2016-04-07    0.001242
Name: ad_created, Length: 76, dtype: float64

For the ad_created column, 2016-04-03 has the highest percentage of 3.88%

In [25]:
#calculating the distribution of last_seen
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001077
2016-03-06    0.004327
2016-03-07    0.005425
2016-03-08    0.007309
2016-03-09    0.009607
2016-03-10    0.010622
2016-03-11    0.012402
2016-03-12    0.023749
2016-03-13    0.008883
2016-03-14    0.012651
2016-03-15    0.015881
2016-03-16    0.016440
2016-03-17    0.028076
2016-03-18    0.007330
2016-03-19    0.015777
2016-03-20    0.020643
2016-03-21    0.020581
2016-03-22    0.021368
2016-03-23    0.018593
2016-03-24    0.019794
2016-03-25    0.019173
2016-03-26    0.016688
2016-03-27    0.015529
2016-03-28    0.020829
2016-03-29    0.022320
2016-03-30    0.024722
2016-03-31    0.023811
2016-04-01    0.022879
2016-04-02    0.024908
2016-04-03    0.025115
2016-04-04    0.024515
2016-04-05    0.125018
2016-04-06    0.221794
2016-04-07    0.132161
Name: last_seen, dtype: float64

For the last_seen column, 2016-04-07 has the highest percentage of 22.17%

In [26]:
autos["registration_year"].describe()

count    48297.000000
mean      2004.724807
std         87.831495
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration_year column has odd numbers like 1000 which was before cars were invented and 9999 which is years into the future.
<br>
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [27]:
autos = autos[autos["registration_year"].between (1910,2016)]


In [28]:
autos["registration_year"].describe()

count    46423.000000
mean      2002.935592
std          7.127037
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [29]:
autos["registration_year"].value_counts(normalize=True)

2000    0.067079
2005    0.062835
1999    0.062060
2004    0.058139
2003    0.058053
          ...   
1938    0.000022
1939    0.000022
1953    0.000022
1943    0.000022
1952    0.000022
Name: registration_year, Length: 78, dtype: float64

After further cleaning and removing any car claimed to be registered before 1910 and after 2016, we have 46,423 rows left.
<br>
Also, most cars were registered in year 2000 with 6.71%

## Exploring Price by Brand

In [30]:
#exploring the unique brand values
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [31]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211296
bmw               0.110053
opel              0.107296
mercedes_benz     0.096590
audi              0.086681
ford              0.069879
renault           0.047110
peugeot           0.029899
fiat              0.025720
seat              0.018267
skoda             0.016436
nissan            0.015337
mazda             0.015273
smart             0.014174
citroen           0.014066
toyota            0.012774
hyundai           0.010017
sonstige_autos    0.009543
volvo             0.009112
mini              0.008789
mitsubishi        0.008186
honda             0.007862
kia               0.007087
alfa_romeo        0.006656
porsche           0.006031
suzuki            0.005924
chevrolet         0.005687
chrysler          0.003533
dacia             0.002650
daihatsu          0.002520
jeep              0.002283
land_rover        0.002111
subaru            0.002111
saab              0.001659
jaguar            0.001551
daewoo            0.001486
trabant           0.001379
r

In [32]:
#selecting the top 5 brands
top_5_brands = autos["brand"].value_counts().index[:5]
print(top_5_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')


There are 40 unique brands in the dataset and Volkswagen has the highest frequency of21.12%.
<br>
We would be aggregating the top 5 brands (Volkswagen, bmw, opel, mercedes_benz, audi)
- Volkswagen 21.12%
- bmw 11.00%
- opel 10.82%
- mercedes_benz 9.65%
- audi 8.67%

###  Aggregating Price Data

In [51]:
top_5_mean_brand_price = {}
for b in top_5_brands:
    selected_brands = autos[autos["brand"] ==b]
    selected_mean_price = selected_brands["price_usd"].mean().round(2)
    top_5_mean_brand_price[b]=selected_mean_price


In [53]:
top_5_mean_brand_price

{'volkswagen': 5431.49,
 'bmw': 8378.43,
 'opel': 2999.63,
 'mercedes_benz': 8664.98,
 'audi': 9376.09}

In [38]:
#sorting the brands according to price in desending order
dict(sorted(mean_top_5_brand_price.items(), key=lambda item:item[1], reverse=True))

{'audi': 9376.088469184891,
 'mercedes_benz': 8664.980151650312,
 'bmw': 8378.434918770796,
 'volkswagen': 5431.494036089306,
 'opel': 2999.6338084721942}

The most expensive brand was audi, followed by Mercedes benz then BMW. Opel was the least expensive of the top 5 brands.

## Data Aggregation by Price and Mileage

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. 

In [54]:
top_5_brand_mean_mileage = {}
for b in top_5_brands:
    selected_brands = autos[autos["brand"]==b]
    mean_mileage = selected_brands["odometer_km"].mean().round(2)
    top_5_brand_mean_mileage[b] = mean_mileage
    
top_5_brand_mean_mileage

{'volkswagen': 128781.73,
 'bmw': 132673.71,
 'opel': 129415.78,
 'mercedes_benz': 130951.16,
 'audi': 129249.5}

The mean mileage for the top 10 brands has been calculated and stored in a dictionary. Now we will convert both dictionaries into series objects, using the series constructor.

In [56]:
#series constructor for top 5 brands mean prices
t5bmp_series = pd.Series(top_5_mean_brand_price)
print(t5bmp_series)

volkswagen       5431.49
bmw              8378.43
opel             2999.63
mercedes_benz    8664.98
audi             9376.09
dtype: float64


In [57]:
#series constructor for top 5 brands mean mileage
t5bmm_series = pd.Series(top_5_brand_mean_mileage)
print(t5bmm_series)

volkswagen       128781.73
bmw              132673.71
opel             129415.78
mercedes_benz    130951.16
audi             129249.50
dtype: float64


We will create a dataframe from the first series object using the dataframe constructor. Then we will assign the other series as a new column in this dataframe

In [58]:
price_miles_df = pd.DataFrame(t5bmp_series, columns=["mean_price"])
price_miles_df ["mean_mileage"] = t5bmm_series

In [60]:
price_miles_df.sort_values('mean_price', ascending=False)

mean_price  mean_mileage
audi              9376.09     129249.50
mercedes_benz     8664.98     130951.16
bmw               8378.43     132673.71
volkswagen        5431.49     128781.73
opel              2999.63     129415.78

In [61]:
price_miles_df.describe()

mean_price  mean_mileage
count     5.000000       5.00000
mean   6970.124000  130214.37600
std    2682.198363    1597.84153
min    2999.630000  128781.73000
25%    5431.490000  129249.50000
50%    8378.430000  129415.78000
75%    8664.980000  130951.16000
max    9376.090000  132673.71000